File: dp/word_related/word_break_ii.rb
# @param {String} s
# @param {String[]} word_dict
# @return {String[]}
def word_break(s, word_dict)
    bt(s, word_dict, {})
end

def bt(s, word_dict, map)
    return map[s] if map.key?(s)
    map[s] = [] if !map[s] # This is the init for a new key

    word_dict.each do |w|
        map[s] << s if w == s # This only happens at the deepest dfs level when we hit an actual word from the dictionary
        next if !s.start_with?(w)
        result_of_rest = bt(s[w.size...s.size], word_dict, map)
        result_of_rest.each do |other_w|
            map[s] << "#{w} #{other_w}"
        end
    end

    map[s]
end

# @param {String} s
# @param {String[]} word_dict
# @return {String[]}
def word_break_dp(s, word_dict)
    word_dict = Set.new(word_dict)
    n = s.size()
    dp = Array.new(n + 1)
    dp[0] = [""]

    1.upto(n) do |i|
        list = []
        0.upto(i - 1) do |j|
            if dp[j] && word_dict.include?(s[j...i])
                dp[j].each do |l|
                    list.push(l + (l.empty? ? "" : " ") + s[j...i])
                end
            end
        end
        dp[i] = list
    end

    dp[n]
end

# 140. Word Break II
# https://leetcode.com/problems/word-break-ii/solution/

# Approach 1: Brute-Force,                                  Time: O(n ^ n), Space: O(n ^ 3)

# Approach 2: Backtracking / Recursion with memoization,    Time: O(n ^ 3), Space: O(n ^ 3)
# Example: s = "catsanddog" word_dict = ["cat","cats","and","sand","dog"]
# bt call with catsanddog {}
# bt call with sanddog {"catsanddog"=>[]}
#     bt call with dog {"catsanddog"=>[], "sanddog"=>[]}
#         bt call with "" {"catsanddog"=>[], "sanddog"=>[], "dog"=>["dog"]}
#         bt return with "" {"catsanddog"=>[], "sanddog"=>[], "dog"=>["dog"], ""=>[]}
#     bt return with dog {"catsanddog"=>[], "sanddog"=>[], "dog"=>["dog"], ""=>[]}
# bt return with sanddog {"catsanddog"=>[], "sanddog"=>["sand dog"], "dog"=>["dog"], ""=>[]}
# bt call with anddog {"catsanddog"=>["cat sand dog"], "sanddog"=>["sand dog"], "dog"=>["dog"], ""=>[]}
#     bt call with dog {"catsanddog"=>["cat sand dog"], "sanddog"=>["sand dog"], "dog"=>["dog"], ""=>[], "anddog"=>[]}
# bt return with anddog {"catsanddog"=>["cat sand dog"], "sanddog"=>["sand dog"], "dog"=>["dog"], ""=>[], "anddog"=>["and dog"]}
# bt return with catsanddog {"catsanddog"=>["cat sand dog", "cats and dog"], "sanddog"=>["sand dog"], "dog"=>["dog"], ""=>[], "anddog"=>["and dog"]}

# bt call with catsanddog
# bt call with sanddog
#     bt call with dog
#         bt call with ""
#         bt return with ""
#     bt return with dog
# bt return with sanddog
# bt call with anddog
#     bt call with dog
# bt return with anddog
# bt return with catsanddog
# Find the recursion tree in the diagram attached.

# Approach 3: DP,                                           Time: O(n ^ 3), Space: O(n ^ 3)
# In the previous approach we can see that many subproblems were redundant,
# i.e we were calling the recursive function multiple times
# for the same substring appearing through multiple paths.
# To avoid this we can use memorization method,
# where we are making use of a hashmap to store the results in the form of a key:value pair.
# In this hashmap, the keykey used is the starting index of the string currently considered and
# the valuevalue contains all the sentences which can be formed
# using the substring from this starting index onwards.
# Thus, if we encounter the same starting index from different function calls,
# we can return the result directly from the hashmap rather than going for redundant function calls.

# With memorization many redundant subproblems are avoided and
# recursion tree is pruned and thus it reduces the time complexity by a large factor.

# Time:   O(n^3) => Size of recursion tree can go up to n^2. The creation of list takes n time.
# Space:  O(n^3) => The depth of the recursion tree can go up to n
#         and each activation record can contains a string list of size n.

require 'set'
require 'test/unit'
extend Test::Unit::Assertions

assert_equal(word_break("catsanddog", ["cat","cats","and","sand","dog"]), ["cat sand dog", "cats and dog"])
assert_equal(word_break("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", ["a","aa","aaa","aaaa","aaaaa","aaaaaa","aaaaaaa","aaaaaaaa","aaaaaaaaa","aaaaaaaaaa"]), [])

assert_equal(word_break_dp("catsanddog", ["cat","cats","and","sand","dog"]), ["cat sand dog", "cats and dog"])
# TLE on below input
# assert_equal(word_break_dp("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", ["a","aa","aaa","aaaa","aaaaa","aaaaaa","aaaaaaa","aaaaaaaa","aaaaaaaaa","aaaaaaaaaa"]), 0)

File: dp/word_related/word_break_ii.py
from typing import List


class Solution1:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        self.memo = {}
        return self.dfs(s, wordDict)

    def dfs(self, s, word_dict):
        if s in self.memo: return self.memo[s]
        if not s: return []

        result = []
        for word in word_dict:
            if not s.startswith(word): continue

            if len(word) == len(s): result.append(word)
            else:
                results_of_rest = self.dfs(s[len(word):], word_dict)
                for item in results_of_rest:
                    item = f"{word} {item}"
                    result.append(item)
        self.memo[s] = result
        return result


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        if not self.is_breakable(s, wordDict): return []
        n, word_dict = len(s), set(wordDict)
        dp = [['']] + [[]] * n

        for i in range(1, n + 1):
            new_list = []
            for j in range(i):
                if not dp[j] or s[j:i] not in word_dict: continue
                new_list.extend([l + ('' if not l else ' ') + s[j:i] for l in dp[j]])
            dp[i] = new_list

        return dp[n]

    def is_breakable(self, s, word_dict):
        word_dict, n = set(word_dict), len(s)
        dp = [True] + [False] * n

        for i in range(1, n + 1):
            for j in range(i):
                if not dp[j] or s[j:i] not in word_dict: continue
                dp[i] = True
                break

        return dp[n]


# 140. Word Break II
# https://leetcode.com/problems/word-break-ii/description/



# Approach 1: DFS with Caching
# This approach is not ideal for interviews because,
# sometimes the interviewer says that the dictionary is not iterable
# and only gives you a API for checking containment

# Steps:
# 1. We iterate through dictionary and and for each word that the input string startswith()
#    we do a DFS to get results of the rest
# 2. We use the results of the rest from the next level of DFS and combine
#    the word with each of those results and return to the prev level

# Complexity Analysis
# Time:
# - Firstly n^2 to get every prefixes
# - Every prefix lets you branch for solving a smaller subproblem of size n - i (i.e n - 1, n - 2, n - 3, ... 1).
# So writing the recurrence for this would be:
# T(n) = n * (1 + T(n - 1) ) + (1 + T(n - 2) ) + ...... T(1)).
#   Here adding the "1" for each subproblem because the entire string is a solution itself,
#   and the factor of n for looping through the possible prefixes.
# PROOF: https://stackoverflow.com/questions/46004100/solving-this-recurrence-without-the-master-theorem-backtracking-algorithm
# This sums up to roughly n * 2^(n-1) which is O(n * 2^n) (Time and memory).
# Space:
# it's same for space as those 2^n can be a valid break, so the space complexity is also O(2^n)

# Time: O(n * 2^n)
# Space: O(n * 2^n)

# Approach 2: DP
# Firstly, check if the word is breakable, and only then do the actual breaking.
# Q: Why does this not take a long time?
# A: Cuz is_breakable() has earrly break for when a possible break is found.

# is_breakable() uses the DP below:
# dp[i] represents the bool of breakability of s[0:i] into words in wordDict
# for every j from 0 to i - 1
#   - dp[i] = True if dp[j] = True and s[j:i] is in the dict
#   - dp[0] = True, empty string is in the dict

# wordBreak() uses the DP below:
# dp[i] represents the actual words you can break s[0:i] from words in wordDict
# for every j from 0 to i - 1:
#   - If dp[j] is not empty and if s[j:i] is in the word_dict, we've found a break point so,
#   - Use the results of s[0:j] and combine those with s[0:j]
#   - Save list of these results as dp[i]

# Time: O(n^3)
# Space: O(n^3)

sol = Solution1()
wordDict = ["cat", "cats", "and", "sand", "dog"]
assert sorted(sol.wordBreak("catsanddog", wordDict)) == sorted([
    "cats and dog",
    "cat sand dog"
])
wordDict = ["apple", "pen", "applepen", "pine", "pineapple"]
assert sorted(sol.wordBreak("pineapplepenapple", wordDict)) == sorted([
    "pine apple pen apple",
    "pineapple pen apple",
    "pine applepen apple"
])
wordDict = ["cats", "dog", "sand", "and", "cat"]
assert sol.wordBreak("catsandog", wordDict) == []
wordDict = [
    "a", "aa", "aaa",
    "aaaa", "aaaaa", "aaaaaa", "aaaaaaa",
    "aaaaaaaa", "aaaaaaaaa", "aaaaaaaaaa"
]
assert sol.wordBreak(
    'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
    wordDict
) == []


sol = Solution()
wordDict = ["cat", "cats", "and", "sand", "dog"]
assert sorted(sol.wordBreak('catsanddog', wordDict)) == sorted([
    "cats and dog",
    "cat sand dog"
])
wordDict = ["apple", "pen", "applepen", "pine", "pineapple"]
assert sorted(sol.wordBreak('pineapplepenapple', wordDict)) == sorted([
    "pine apple pen apple",
    "pineapple pen apple",
    "pine applepen apple"
])
wordDict = ["cats", "dog", "sand", "and", "cat"]
assert sol.wordBreak('catsandog', wordDict) == []
wordDict = [
    "a", "aa", "aaa",
    "aaaa", "aaaaa", "aaaaaa", "aaaaaaa",
    "aaaaaaaa", "aaaaaaaaa", "aaaaaaaaaa"
]
assert sol.wordBreak(
    'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaabaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
    wordDict
) == []
